x-pack/plugins/enterprise_search/SERVER.md

```diff
@@ -0,0 +1,78 @@
+
+Where necessary to make these functions readable and/or avoid duplicate code, these functions should call utility functions located in a `shared` folder in the nearest common parent directory. Eg a library function called by routes in just the `indices` directory should be located in [server/lib/indices/shared](server/lib/indices/shared), while a library function called by routes in both the `connectors` and `indices` directories should be located in [server/lib/shared](server/lib/shared).
+
+Be careful when sharing functions across multiple routes: if you're adding many inputs to a single function, it's probably better to split them up and dedicate each to a single route, even if that means more duplicate code in the system. The added complexity caused by maintaining multiple code paths in a single function is generally not worth it.
+
+### Types and validation
+
+Endpoints validate their inputs using [@kbn/config-schema](/packages/kbn-config-schema/), and we share TypeScript types between the frontend and backend in [common/types](common/types) to facilitate consistent API expectation. A few hints:
+
+- schema.maybe(...) allows the `...` to be optional/`undefined`.
+- schema.nullable(...) allows the `...` to be `null`, which is distinct from making it optional or allowing `undefined`.
+- Lean on `elasticsearch-js` (the built-in Elasticsearch client) types where possible to do the heavy lifting.
+
+### Working with Elasticsearch and elasticsearch-js
+
+Any Elasticsearch function should be available via the client, but it can be a bit hard to figure out what the required function name is. Because the (elasticsearch-js)[elasticsearch-js] documentation isn't great, a search through Kibana's codebase is often more efficient than trying to go through the docs.
+
+The Kibana plugin provides an Elasticsearch client object to every route. You can access this client under the obvious name `client`. That client comes with two users you can use to perform actual Elasticsearch operations: `client.asCurrentUser` and `client.asInternalUser`. `asCurrentUser` will execute operations using the permissions attached to the incoming request, which should be the permissions of the user firing that request in Kibana. `asInternalUser` will execute requests as Kibana's internal user. This user has very limited permissions and is generally only useful if you want to manipulate Kibana's Saved Objects.
+
+For most operations you'll want to specify a generic type argument to indicate the expected return type, as TypeScript has no way of knowing that type. See:
+
+```
+const connectorResult = await client.asCurrentUser.search<T>({
```

Comment:
The general answer is of course it depends, but in most cases you're not returning the full Elasticsearch response but only the inner document (or some part of it). So yes, in that case you'd just return the inner license property, not the full response. Do you think that needs clarification here?

Reply:
I'm not sure what you mean by standardizing those types. You mean the ones that go to the Rails app? In what way would you standardize them?

Reply:
Standarize as in giving some guidelines on what to do with types. i.e. the example I gave above and similar rules(I cannot give some concrete examples yet). I think this could be iteration over this PR, so we are good for now.